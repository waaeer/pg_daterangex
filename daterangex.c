#include "postgres.h"
#include "utils/rangetypes.h"
#include "utils/date.h"
#include "utils/fmgrprotos.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(daterangex_canonical);

Datum
daterangex_canonical(PG_FUNCTION_ARGS)
{
	RangeType  *r = PG_GETARG_RANGE_P(0);
	Node	   *escontext = fcinfo->context;
	TypeCacheEntry *typcache;
	RangeBound	lower;
	RangeBound	upper;
	bool		empty;

	typcache = range_get_typcache(fcinfo, RangeTypeGetOid(r));

	range_deserialize(typcache, r, &lower, &upper, &empty);

	if (empty)
		PG_RETURN_RANGE_P(r);

	if (!lower.infinite && !DATE_NOT_FINITE(DatumGetDateADT(lower.val)) &&
		!lower.inclusive)
	{
		DateADT		bnd = DatumGetDateADT(lower.val);

		bnd++;
		if (unlikely(!IS_VALID_DATE(bnd)))
			ereturn(escontext, (Datum) 0,
					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
					 errmsg("date out of range")));
		lower.val = DateADTGetDatum(bnd);
		lower.inclusive = true;
	}

	if (!upper.infinite && !DATE_NOT_FINITE(DatumGetDateADT(upper.val)) &&
		!upper.inclusive)
	{
		DateADT		bnd = DatumGetDateADT(upper.val);

		bnd--;
		if (unlikely(!IS_VALID_DATE(bnd)))
			ereturn(escontext, (Datum) 0,
					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
					 errmsg("date out of range")));
		upper.val = DateADTGetDatum(bnd);
		upper.inclusive = true;
	}

	PG_RETURN_RANGE_P(range_serialize(typcache, &lower, &upper,
									  false, escontext));
}

PG_FUNCTION_INFO_V1(daterangex_to_daterange);

Datum
daterangex_to_daterange(PG_FUNCTION_ARGS)
{
	return daterange_canonical(fcinfo);
}

PG_FUNCTION_INFO_V1(daterange_to_daterangex);

Datum
daterange_to_daterangex(PG_FUNCTION_ARGS)
{	
	RangeType  *r = PG_GETARG_RANGE_P(0);
	RangeTypeGetOid(r) = 2246043;
	return daterangex_canonical(fcinfo);
}




