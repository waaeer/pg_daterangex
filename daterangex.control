# daterangex extension
comment = 'adterange with inclusive upper bound'
default_version = '1.0'
module_pathname = '$libdir/daterangex'
relocatable = true
trusted = true

