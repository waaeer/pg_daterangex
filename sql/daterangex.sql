CREATE EXTENSION daterangex;
SELECT daterangex('2024-01-01' , '2024-06-01', '[]');
SELECT daterangex('2024-01-01' , '2024-06-01', '[)');
SELECT '[2025-01-01,2025-02-02]'::daterange::daterangex;
SELECT '[2025-01-01,2025-02-02)'::daterange::daterangex;
SELECT '[2025-01-01,2025-02-02]'::daterangex::daterange;
SELECT '[2025-01-01,2025-02-02)'::daterangex::daterange;

CREATE TABLE x (y daterangex);
INSERT INTO x SELECT daterangex(d::date, (d + '1month'::interval)::date, '[]') FROM generate_series('2015-01-01'::date,'2025-02-02', '1day') d;
INSERT INTO x SELECT daterangex(d::date, NULL::date, '[)') FROM generate_series('2015-01-01'::date,'2025-02-02', '1day') d;
SELECT COUNT(*) FROM x;
CREATE INDEX ON x USING gist (y);
INSERT INTO x SELECT '[2024-01-01,)'::daterange::daterangex;



