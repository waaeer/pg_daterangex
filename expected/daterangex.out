CREATE EXTENSION daterangex;
SELECT daterangex('2024-01-01' , '2024-06-01', '[]');
       daterangex        
-------------------------
 [01-01-2024,06-01-2024]
(1 row)

SELECT daterangex('2024-01-01' , '2024-06-01', '[)');
       daterangex        
-------------------------
 [01-01-2024,05-31-2024]
(1 row)

SELECT '[2025-01-01,2025-02-02]'::daterange::daterangex;
       daterangex        
-------------------------
 [01-01-2025,02-02-2025]
(1 row)

SELECT '[2025-01-01,2025-02-02)'::daterange::daterangex;
       daterangex        
-------------------------
 [01-01-2025,02-01-2025]
(1 row)

SELECT '[2025-01-01,2025-02-02]'::daterangex::daterange;
        daterange        
-------------------------
 [01-01-2025,02-03-2025)
(1 row)

SELECT '[2025-01-01,2025-02-02)'::daterangex::daterange;
        daterange        
-------------------------
 [01-01-2025,02-02-2025)
(1 row)

CREATE TABLE x (y daterangex);
INSERT INTO x SELECT daterangex(d::date, (d + '1month'::interval)::date, '[]') FROM generate_series('2015-01-01'::date,'2025-02-02', '1day') d;
INSERT INTO x SELECT daterangex(d::date, NULL::date, '[)') FROM generate_series('2015-01-01'::date,'2025-02-02', '1day') d;
SELECT COUNT(*) FROM x;
 count 
-------
  7372
(1 row)

CREATE INDEX ON x USING gist (y);
INSERT INTO x SELECT '[2024-01-01,)'::daterange::daterangex;
